JavaFiddle-Web/src/main/java/ru/javafiddle/web/services/ProjectService.java

```diff
@@ -70,22 +100,44 @@ public Response getProjectStructure(@PathParam("projectHash") String projectHash
         return Response.ok(projectTree).build();
     }
 
+
     @POST
     @Consumes(MediaType.APPLICATION_JSON)
     public Response createProject(ProjectInfo projectInfo, @Context UriInfo uriInfo) {
 
         try {
-            String projectHash = "INVALID PROJECT HASH";
-//!TODO
-//            projectHash = projectBean.createProject(projectInfo.getUserNickName(),
-//                    projectInfo.getProjectHash(),
-//                    projectInfo.getProjectName());
-
-            URI uri = uriInfo.getAbsolutePathBuilder().path(projectHash).build();
-            return Response.created(uri).build();
+            //get Group
+            Integer groupId = projectInfo.getGroupId();
+            Group group = (groupId == null) ? getDefaultGroup() : groupBean.getGroupByGroupId(groupId);
+
+            //if hash field is not null then clone the project
+            String hash = projectInfo.getProjectHash();
+            if (hash != null) {
+                Project project = projectBean.getProjectByProjectHash(hash);
+                if (project == null) { //if specified hash does not exist return NOT_FOUND status
+                    return Response.status(Response.Status.NOT_FOUND).build();
+                }
+                project.setGroup(group); //set specified group
+                cloneProject(project);
+                return Response.ok().build();
+            }
+
+            //try to create new project with given name
+            String projectName = projectInfo.getProjectName();
+
+            if (projectName != null) {
+
+                Project project = new Project(projectName, group);
+                project = projectBean.createProject(project);
+                createProjectFile(project);
+                return Response.ok().build();
+            }
+
+            // if both name and hash are null it is impossible to create project
+            return Response.status(Response.Status.BAD_REQUEST).build();
 
-        } catch(Exception e){
-            return Response.serverError().build();
+        } catch (UnsupportedEncodingException | NoSuchAlgorithmException e) {
```

Comment:
@alexander1leonov , I don't know how to minimise it, couse there is a couple of methods inside it, which throw such exceptions. If it wasn't, I would cover only one and try/catch block would be smaller, but here, method will become bigger. Is it worth doing that?